BusinessLayer/SongControl/SongControl.cpp

```diff
@@ -16,24 +20,42 @@ SongControl::~SongControl()
 QString SongControl::nextSong()
 {
     currentSongIndex_ = (currentSongIndex_ + 1) % files_.size();
+    playOrder_.prepend(currentSongIndex_);
+    playOrderIndex_ = 0;
     return files_[currentSongIndex_];
 }
 
 QString SongControl::previousSong()
 {
-    currentSongIndex_ -= 1;
-    if (currentSongIndex_ < 0)
+
+    if (playOrderIndex_ < 9)
     {
-        currentSongIndex_ = files_.size() - 1;
+        playOrderIndex_ += 1;
     }
-    return files_[currentSongIndex_];
+
+    if (playOrder_[playOrderIndex_] == -1)
+    {
+        return files_[currentSongIndex_];
+    }
+
+    return files_[playOrder_[playOrderIndex_]];
 }
 
 QString SongControl::currentSong()
 {
+    playOrder_.prepend(currentSongIndex_);
+    playOrderIndex_ = 0;
     return (files_[(currentSongIndex_)]);
 }
 
+QString SongControl::shuffleSong()
+{
+    currentSongIndex_ = (rand() % files_.size());
```

Comment:
might want to add a check that the number generated is not the number currently playing. During a quick test of the program I had the same song play 3 times in a row in a test library of about 35 songs.